File: piflow-core/src/main/scala/cn/piflow/path.scala
package cn.piflow

import scala.collection.mutable.{ArrayBuffer, Map => MMap}

trait PathHead {
  def via(ports: (String, String)): PathVia;

  def to(stopTo: String): Path;
}

trait Path extends PathHead {
  def toEdges(): Seq[Edge];

  def addEdge(edge: Edge): Path;
}

trait PathVia {
  def to(stopTo: String): Path;
}

class PathImpl() extends Path {
  val edges = ArrayBuffer[Edge]();

  override def toEdges(): Seq[Edge] = edges.toSeq;

  override def addEdge(edge: Edge): Path = {
    edges += edge;
    this;
  }

  val thisPath = this;

  override def via(ports: (String, String)): PathVia = new PathVia() {
    override def to(stopTo: String): Path = {
      edges += new Edge(edges.last.stopTo, stopTo, ports._1, ports._2);
      thisPath;
    }
  }

  override def to(stopTo: String): Path = {
    edges += new Edge(edges.last.stopTo, stopTo, "", "");
    this;
  }
}

case class Edge(stopFrom: String, stopTo: String, outport: String, inport: String) {
  override def toString() = {
    s"[$stopFrom]-($outport)-($inport)-[$stopTo]";
  }
}

object Path {
  def from(stopFrom: String): PathHead = {
    new PathHead() {
      override def via(ports: (String, String)): PathVia = new PathVia() {
        override def to(stopTo: String): Path = {
          val path = new PathImpl();
          path.addEdge(new Edge(stopFrom, stopTo, ports._1, ports._2));
          path;
        }
      }

      override def to(stopTo: String): Path = {
        val path = new PathImpl();
        path.addEdge(new Edge(stopFrom, stopTo, "", ""));
        path;
      }
    };
  }

  def of(path: (Any, String)): Path = {
    val pi = new PathImpl();

    def _addEdges(path: (Any, String)): Unit = {
      val value1 = path._1;

      //String->String
      if (value1.isInstanceOf[String]) {
        pi.addEdge(new Edge(value1.asInstanceOf[String], path._2, "", ""));
      }

      //(String->String)->String
      else if (value1.isInstanceOf[(Any, String)]) {
        val tuple = value1.asInstanceOf[(Any, String)];
        _addEdges(tuple);
        pi.addEdge(new Edge(tuple._2, path._2, "", ""));
      }

      else {
        throw new InvalidPathException(value1);
      }
    }

    _addEdges(path);
    pi;
  }
}

File: piflow_engine/cn/piflow/core/path.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


class InvalidPathError(ValueError):
    pass


@dataclass
class Edge:
    stop_from: str
    stop_to: str
    out_port: str = ""
    in_port: str = ""

    def __str__(self) -> str:
        return f"[{self.stop_from}]-({self.out_port})-({self.in_port})-[{self.stop_to}]"


class PathVia:
    def __init__(self, path: "Path", out_port: str, in_port: str):
        self._path = path
        self._out_port = out_port
        self._in_port = in_port

    def to(self, stop_to: str) -> "Path":
        if not self._path.edges:
            raise InvalidPathError("path must contain at least one edge before via().to()")
        last_stop = self._path.edges[-1].stop_to
        self._path.add_edge(Edge(last_stop, stop_to, self._out_port, self._in_port))
        return self._path


@dataclass
class Path:
    edges: list[Edge] = field(default_factory=list)

    def to_edges(self) -> list[Edge]:
        return list(self.edges)

    def add_edge(self, edge: Edge) -> "Path":
        self.edges.append(edge)
        return self

    def via(self, outport: str, inport: str) -> PathVia:
        return PathVia(self, outport, inport)

    def to(self, stop_to: str) -> "Path":
        if not self.edges:
            raise InvalidPathError("path must contain at least one edge before to()")
        self.edges.append(Edge(self.edges[-1].stop_to, stop_to, "", ""))
        return self

    @classmethod
    def from_(cls, stop_from: str) -> "_PathHead":
        return _PathHead(stop_from)

    @classmethod
    def of(cls, path: tuple[Any, str]) -> "Path":
        result = cls()

        def _add_edges(item: tuple[Any, str]) -> None:
            value1, value2 = item
            if isinstance(value1, str):
                result.add_edge(Edge(value1, value2, "", ""))
            elif isinstance(value1, tuple) and len(value1) == 2:
                _add_edges(value1)
                result.add_edge(Edge(value1[1], value2, "", ""))
            else:
                raise InvalidPathError(f"invalid path head: {value1}")

        _add_edges(path)
        return result


class _PathHead:
    def __init__(self, stop_from: str):
        self._stop_from = stop_from

    def via(self, outport: str, inport: str) -> PathVia:
        path = Path()
        path.add_edge(Edge(self._stop_from, "", outport, inport))
        return _PathHeadVia(path)

    def to(self, stop_to: str) -> Path:
        path = Path()
        path.add_edge(Edge(self._stop_from, stop_to, "", ""))
        return path


class _PathHeadVia(PathVia):
    def __init__(self, path: Path):
        self._path = path

    def to(self, stop_to: str) -> Path:
        if not self._path.edges:
            raise InvalidPathError("path must contain at least one edge before via().to()")
        first_edge = self._path.edges[0]
        first_edge.stop_to = stop_to
        return self._path
